### PR TITLE
[RHTAPBUGS-418]multiple components with context path

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Test for [RHTAPBUGS-418](https://issues.redhat.com//browse/RHTAPBUGS-418)

### What does this PR do?:
Test for an issue I saw in the UI where importing a project with multiple components and a context path set results in the context path going missing.

I could not get this test to run locally as all the detection tests are timing out for me.

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/browse/RHTAPBUGS-418

### CODE REVIEW

File: pkg/devfile/devfile.go [hunk 1]
```diff
@@ -851,7 +851,28 @@ func DownloadDevfileAndDockerfile(url string) ([]byte, string, []byte, string) {
 // Map 3 returns a context to the Dockerfile uri or a matched DockerfileURL from the devfile registry if no Dockerfile/Containerfile is present in the context
 // Map 4 returns a context to the list of ports that were detected by alizer in the source code, at that given context
 func ScanRepo(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, map[string][]int, error) {
```
Comment (stuartwdouglas) [inline pkg/devfile/devfile.go:853]: I am not sure if the original method could still be used anywhere else, so I added a new one instead

Comment (stuartwdouglas) [reply pkg/devfile/devfile.go]: I have updated the PR

**Comment by stuartwdouglas** [comment]
Does this need any more action on my part?

File: controllers/componentdetectionquery_controller.go [hunk 2]
```diff
@@ -284,7 +284,7 @@ func (r *ComponentDetectionQueryReconciler) Reconcile(ctx context.Context, req c
 			if isMultiComponent {
 				log.Info(fmt.Sprintf("Since this is a multi-component, attempt will be made to read only level 1 dir for devfiles... %v", req.NamespacedName))
 
-				devfilesMap, devfilesURLMap, dockerfileContextMap, componentPortsMap, err = devfile.ScanRepo(log, r.AlizerClient, componentPath, r.DevfileRegistryURL, source)
+				devfilesMap, devfilesURLMap, dockerfileContextMap, componentPortsMap, err = devfile.ScanRepo(log, r.AlizerClient, source.Context, componentPath, r.DevfileRegistryURL, source)
```
Comment (yangcao77) [inline controllers/componentdetectionquery_controller.go:287]: the source is already been passed in as a arg. no need an extra arg for `source.Context`

Comment (stuartwdouglas) [reply controllers/componentdetectionquery_controller.go:287]: oops, fixed

File: pkg/devfile/devfile.go [hunk 3]
```diff
@@ -851,7 +853,15 @@ func DownloadDevfileAndDockerfile(url string) ([]byte, string, []byte, string) {
...
-	return search(log, a, localpath, devfileRegistryURL, source)
+	originalContext := source.Context
```
Comment (yangcao77) [inline pkg/devfile/devfile.go:856]: can this block be moved to `search`?  
just trying to avoid the API arg change.

**Review by yangcao77 — CHANGES_REQUESTED** [review]

File: pkg/devfile/devfile.go (hunk 3 above)
Comment (yangcao77) [reply pkg/devfile/devfile.go:856]: actually, this code block can be removed. the `source.context` can be get in `search`

File: pkg/devfile/detect.go [hunk 4]
```diff
@@ -63,7 +63,8 @@ func search(log logr.Logger, a Alizer, localpath string, devfileRegistryURL stri
 			isDevfilePresent := false
 			isDockerfilePresent := false
 			curPath := path.Join(localpath, f.Name())
-			context := f.Name()
+			dirName := f.Name()
+			context := originalContext + f.Name()
```
Comment (yangcao77) [inline pkg/devfile/detect.go:67]: context := path.Join(source.Context, f.Name())

File: pkg/devfile/devfile.go (hunk 3 above)
Comment (yangcao77) [reply pkg/devfile/devfile.go:856]: see https://github.com/redhat-appstudio/application-service/pull/346/files#r1255764499

**Comment by stuartwdouglas** [comment]
Fixed

File: pkg/devfile/detect.go [hunk 5]
```diff
@@ -47,7 +47,14 @@ type AlizerClient struct {
...
 func search(log logr.Logger, a Alizer, localpath string, devfileRegistryURL string, source appstudiov1alpha1.GitSource) (map[string][]byte, map[string]string, map[string]string, map[string][]int, error) {
-
+	originalContext := source.Context
+	if originalContext != "" {
+		if originalContext == "./" {
+			originalContext = ""
+		} else if !strings.HasSuffix(originalContext, "/") {
+			originalContext += "/"
+		}
+	}
```
Comment (yangcao77) [inline pkg/devfile/detect.go:57]: do not need this code change.  the context will never be empty. as it's already set in https://github.com/redhat-appstudio/application-service/blob/eeef42302b0bfcac2877973878966f0e880b7fe7/controllers/componentdetectionquery_controller.go#L147-L149

and with the `path.join`, you do not need to append `/` at the end

**Review by yangcao77 — CHANGES_REQUESTED** [review]

**Review by yangcao77 — APPROVED** [review]

**Pull request merged**